src/parser.ts

```diff
@@ -13,6 +13,21 @@ let PlaceHolders = {
     ArrowParameterPlaceHolder: 'ArrowParameterPlaceHolder'
 };
 
+function upgradeToken(token) {
```

Comment:
An alternative is to make upgradeToken a method of the scanner that receives the strict mode state as an argument. It reads like feature envy in its current form. This function cares about the token, it's type, it's value, its status as a strict mode reserved word, and whether or not the current state is strict. 4 out of 5 of those things are responsibilities of the scanner.